### PR TITLE
Fix queries and storage block operations synchronisation

### PR DESCRIPTION
**Pull request by kolesnikovae**
Resolves https://github.com/grafana/phlare/issues/687

### CODE REVIEW

**Comment by kolesnikovae** [comment]
In addition, testing revealed one more bug I can easily reproduce on `main` (`d3033486`) – fixed as well

```
panic serving 192.168.1.64:63941: runtime error: index out of range [7] with length 0
golang.org/x/net/http2.(*serverConn).runHandler.func1()
        /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2304 +0x20c
panic({0x1067e4f00, 0xc00b27e210})
        /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/runtime/panic.go:890 +0x248
github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5.1()
        /Users/kolesnikovae/go/pkg/mod/github.com/opentracing-contrib/go-stdlib@v1.0.0/nethttp/server.go:150 +0x19c
panic({0x1067e4f00, 0xc00b27e210})
        /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/runtime/panic.go:890 +0x248
github.com/grafana/phlare/pkg/phlaredb.(*headOnDiskQuerier).rowGroup(0xc00766ca90)
        /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/head_queriers.go:29 +0x194
github.com/grafana/phlare/pkg/phlaredb.(*headOnDiskQuerier).SelectMatchingProfiles(0xc00766ca90, {0x106b92d78, 0xc00739e050}, 0xc00739e000)
        /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/head_queriers.go:51 +0x18c
github.com/grafana/phlare/pkg/phlaredb.Queriers.MergeProfilesLabels({0xc00485ee00, 0x33, 0x33}, {0x106b92e20, 0xc00111e870}, 0x10000000?)
        /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/block_querier.go:617 +0xdc8
```

**Comment by kolesnikovae** [comment]
A data race spotted by race detector – fixed:
```
  github.com/prometheus/common/model.(*Time).MarshalJSON()
      <autogenerated>:1 +0x34
  encoding/json.addrMarshalerEncoder()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:494 +0x17c
  encoding/json.condAddrEncoder.encode()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:958 +0x78
  encoding/json.condAddrEncoder.encode-fm()
      <autogenerated>:1 +0x78
  encoding/json.structEncoder.encode()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:759 +0x1bc
  encoding/json.structEncoder.encode-fm()
      <autogenerated>:1 +0x94
  encoding/json.ptrEncoder.encode()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:943 +0x2a4
  encoding/json.ptrEncoder.encode-fm()
      <autogenerated>:1 +0x6c
  encoding/json.(*encodeState).reflectValue()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:358 +0x74
  encoding/json.(*encodeState).marshal()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:330 +0x1a0
  encoding/json.Marshal()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:161 +0xa0
  encoding/json.MarshalIndent()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/encoding/json/encode.go:174 +0x44
  github.com/  github.com/grafana/phlare/pkg/phlaredb.(*Head).flush()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/head.go:940 +0xdec
  github.com/grafana/phlare/pkg/phlaredb.(*Head).Flush()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/head.go:874 +0x8c
  github.com/grafana/phlare/pkg/phlaredb.(*PhlareDB).Flush()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/phlaredb.go:343 +0x44
  github.com/grafana/phlare/pkg/phlaredb.(*PhlareDB).loop()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/phlaredb.go:148 +0x1b8
  github.com/grafana/phlare/pkg/phlaredb.New.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/phlaredb.go:106 +0x34

Previous write at 0x00c004597ae8 by goroutine 50967:
  github.com/grafana/phlare/pkg/phlaredb.(*Head).Ingest()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/head.go:416 +0xe2c
  github.com/grafana/phlare/pkg/ingester.(*Ingester).Push.func1.1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:225 +0x128
  github.com/grafana/phlare/pkg/pprof.FromBytes()
      /Users/kolesnikovae/src/grafana/phlare/pkg/pprof/pprof.go:117 +0x14c
  github.com/grafana/phlare/pkg/ingester.(*Ingester).Push.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:220 +0x428
  github.com/grafana/phlare/pkg/ingester.forInstanceUnary[...].func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:174 +0x40
  github.com/grafana/phlare/pkg/ingester.(*Ingester).forInstance()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:193 +0x214
  github.com/grafana/phlare/pkg/ingester.forInstanceUnary[...]()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:173 +0x74
  github.com/grafana/phlare/pkg/ingester.(*Ingester).Push()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:216 +0x6c
  github.com/grafana/phlare/api/gen/proto/go/ingester/v1/ingesterv1connect.IngesterServiceHandler.Push-fm()
      <autogenerated>:1 +0x60
  gi  github.com/grafana/phlare/pkg/tenant.(*authInterceptor).WrapUnary.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/tenant/interceptor.go:47 +0x36c
  github.com/bufbuild/connect-go.NewUnaryHandler[...].func2()
      /Users/kolesnikovae/go/pkg/mod/github.com/bufbuild/connect-go@v1.4.1/handler.go:75 +0x274
  github.com/bufbuild/connect-go.(*Handler).ServeHTTP()
      /Users/kolesnikovae/go/pkg/mod/github.com/bufbuild/connect-go@v1.4.1/handler.go:229 +0x710
  github.com/gorilla/mux.(*Router).ServeHTTP()
      /Users/kolesnikovae/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x298
  github.com/weaveworks/common/middleware.Instrument.Wrap.func1.2()
      /Users/kolesnikovae/go/pkg/mod/github.com/weaveworks/common@v0.0.0-20221201103051-7c2720a9024d/middleware/instrument.go:70 +0x5c
  github.com/felixge/httpsnoop.(*Metrics).CaptureMetrics()
      /Users/kolesnikovae/go/pkg/mod/github.com/felixge/httpsnoop@v1.0.3/capture_metrics.go:84 +0x290
  github.com/felixge/httpsnoop.CaptureMetricsFn()
      /Users/kolesnikovae/go/pkg/mod/github.com/felixge/httpsnoop@v1.0.3/capture_metrics.go:39 +0x380
  github.com/weaveworks/common/middleware.Instrument.Wrap.func1()
      /Users/kolesnikovae/go/pkg/mod/github.com/weaveworks/common@v0.0.0-20221201103051-7c2720a9024d/middleware/instrument.go:69 +0x318
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  github.com/grafana/phlare/pkg/util.Log.Wrap.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/util/http.go:165 +0x574
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  net/http.Handler.ServeHTTP-fm()
      <autogenerated>:1 +0x60
  github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5()
      /Users/kolesnikovae/go/pkg/mod/github.com/opentracing-contrib/go-stdlib@v1.0.0/nethttp/server.go:154 +0x510
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  net/http.Handler.ServeHTTP-fm()
      <autogenerated>:1 +0x60
  golang.org/x/net/http2.(*serverConn).runHandler()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2311 +0xa4
  golang.org/x/net/http2.(*serverConn).processHeaders.func1()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2025 +0x58

Goroutine 545 (running) created at:
  github.com/grafana/phlare/pkg/phlaredb.New()
      /Users/kolesnikovae/src/grafana/phlare/pkg/phlaredb/phlaredb.go:106 +0x4c4
  github.com/grafana/phlare/pkg/ingester.newInstance()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/instance.go:36 +0x10c
  github.com/grafana/phlare/pkg/ingester.(*Ingester).GetOrCreateInstance()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:152 +0x1e0
  github.com/grafana/phlare/pkg/ingester.(*Ingester).forInstance()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/ingester.go:189 +0x144
  github.com/grafana/phlare/pkg/ingester.(*Ingester).MergeProfilesLabels()
      /Users/kolesnikovae/src/grafana/phlare/pkg/ingester/query.go:47 +0x68
  github.com/grafana/phlare/api/gen/proto/go/ingester/v1/ingesterv1connect.IngesterServiceHandler.MergeProfilesLabels-fm()
      <autogenerated>:1 +0x60
  github.com/bufbuild/connect-go.NewBidiStreamHandler[...].func1()
      /Users/kolesnikovae/go/pkg/mod/github.com/bufbuild/connect-go@v1.4.1/handler.go:160 +0x9c
  github.com/grafana/phlare/pkg/tenant.(*authInterceptor).WrapStreamingHandler.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/tenant/interceptor.go:73 +0x1f4
  github.com/bufbuild/connect-go.(*Handler).ServeHTTP()
      /Users/kolesnikovae/go/pkg/mod/github.com/bufbuild/connect-go@v1.4.1/handler.go:229 +0x710
  github.com/gorilla/mux.(*Router).ServeHTTP()
      /Users/kolesnikovae/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x298
  github.com/weaveworks/common/middleware.Instrument.Wrap.func1.2()
      /Users/kolesnikovae/go/pkg/mod/github.com/weaveworks/common@v0.0.0-20221201103051-7c2720a9024d/middleware/instrument.go:70 +0x5c
  github.com/felixge/httpsnoop.(*Metrics).CaptureMetrics()
      /Users/kolesnikovae/go/pkg/mod/github.com/felixge/httpsnoop@v1.0.3/capture_metrics.go:84 +0x290
  github.com/felixge/httpsnoop.CaptureMetricsFn()
      /Users/kolesnikovae/go/pkg/mod/github.com/felixge/httpsnoop@v1.0.3/capture_metrics.go:39 +0x380
  github.com/weaveworks/common/middleware.Instrument.Wrap.func1()
      /Users/kolesnikovae/go/pkg/mod/github.com/weaveworks/common@v0.0.0-20221201103051-7c2720a9024d/middleware/instrument.go:69 +0x318
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  github.com/grafana/phlare/pkg/util.Log.Wrap.func1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/util/http.go:165 +0x574
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  net/http.Handler.ServeHTTP-fm()
      <autogenerated>:1 +  github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5()
      /Users/kolesnikovae/go/pkg/mod/github.com/opentracing-contrib/go-stdlib@v1.0.0/nethttp/server.go:154 +0x510
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  net/http.Handler.ServeHTTP-fm()
      <autogenerated>:1 +0x60
  golang.org/x/net/http2.(*serverConn).runHandler()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2311 +0xa4
  golang.org/x/net/http2.(*serverConn).processHeaders.func1()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2025 +0x58

Goroutine 50967 (running) created at:
  golang.org/x/net/http2.(*serverConn).processHeaders()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:2025 +0x9e4
  golang.org/x/net/http2.(*serverConn).processFrame()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:1530 +0x320
  golang.org/x/net/http2.(*serverConn).processFrameFromReader()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:1473 +0x264
  golang.org/x/net/http2.(*serverConn).serve()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:957 +0xfcc
  golang.org/x/net/http2.(*Server).ServeConn()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/server.go:531 +0x10d0
  golang.org/x/net/http2/h2c.h2cHandler.ServeHTTP()
      /Users/kolesnikovae/go/pkg/mod/golang.org/x/net@v0.8.0/http2/h2c/h2c.go:97 +0x584
  golang.org/x/net/http2/h2c.(*h2cHandler).ServeHTTP()
      <autogenerated>:1 +0x64
  github.com/grafana/phlare/pkg/util.glob..func2.1()
      /Users/kolesnikovae/src/grafana/phlare/pkg/util/recovery.go:32 +0x94
  net/http.HandlerFunc.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2122 +0x48
  net/http.serverHandler.ServeHTTP()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:2936 +0x530
  net/http.(*conn).serve()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:1995 +0x898
  net/http.(*Server).Serve.func3()
      /Users/kolesnikovae/.gvm/gos/go1.20rc3/src/net/http/server.go:3089 +0x4c
==================
```

**Comment by kolesnikovae** [comment]
I'm going to leave my stress running over night, to ensure the fix does not introduce new bugs:

50 clients with "real" profiles, 5 parallel queries (last hour), phlare is compiled with the race detector enabled and is running with these options:
```
-phlaredb.max-block-duration=1m -phlaredb.row-group-target-size=5000000 -distributor.push.timeout=15s
```

File: pkg/phlaredb/deduplicating_slice.go [hunk 1]
```diff
@@ -110,17 +109,14 @@ func (s *deduplicatingSlice[M, K, H, P]) maxRowsPerRowGroup() int {
...
-		s.buffer = parquet.NewBuffer(
-			s.persister.Schema(),
-			parquet.SortingRowGroupConfig(s.persister.SortingColumns()),
-			parquet.ColumnBufferCapacity(s.cfg.MaxBufferRowCount),
-		)
-	}
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	buffer := parquet.NewBuffer(
```
Comment (cyriltovena) [inline pkg/phlaredb/deduplicating_slice.go:115]: Not possible to re-use it anymore ?

**Review by cyriltovena — APPROVED** [review]
Nice work, I'd love to see how this impact our tail latency.

LGTM 👏

File: pkg/phlaredb/deduplicating_slice.go (hunk 1 above)
Comment (kolesnikovae) [reply pkg/phlaredb/deduplicating_slice.go:115]: I'm afraid I failed to find where we reuse dedup slice or its parts: we use zero values at head initialization: https://github.com/grafana/phlare/blob/d3033486455ae74ff4c5938eaaa4c2ff15b39240/pkg/phlaredb/head.go#L158 

and the `Init` function does not seem to try to reuse allocated objects by any means:
https://github.com/grafana/phlare/blob/d3033486455ae74ff4c5938eaaa4c2ff15b39240/pkg/phlaredb/deduplicating_slice.go#L61-L77

The buffer itself is not used anywhere but at `Flush`.

I should say that if we want to flush the head without blocking writes to the db, we have to sacrifice some memory. Although a pool could help there, judging by the GC rate, it's very likely the objects will be removed from it before the next flush. My conclusion is, provided that I don't find this allocation particularly harmful (it is almost unnoticeable in the heap profile), performance improvement would be small to unnoticeable. Therefore I'm not very concerned about this

**Comment by kolesnikovae** [comment]
Also fixed one more bug with synchronisation of the block eviction and the block querier synchronisation – turns out [the recent change](https://github.com/grafana/phlare/pull/628) didn't not fully eliminate the race: 
image

**Pull request merged**